sno/dataset2.py

```diff
@@ -0,0 +1,218 @@
+        return json.dumps(self.all_columns).encode("utf8")
+
+    def merge_pk_and_non_pk(self, pk_values, non_pk_values):
+        """
+        Given the pk_values (in pk_index order) and the non_pk_values (in row order),
+        merge them together to get a tuple of all values (in row order).
+        """
+        assert len(pk_values) == len(self.pk_columns)
+        assert len(non_pk_values) == len(self.non_pk_columns)
+
+        iter_non_pk_values = iter(non_pk_values)
+        all_values = []
+        for column in self.all_columns:
+            pk_index = _pk_index(column)
+            if pk_index is None:
+                all_values.append(next(iter_non_pk_values))
+            else:
+                all_values.append(pk_values[pk_index])
+        return tuple(all_values)
+
+    def split_pk_and_non_pk(self, all_values):
+        """
+        Given all the row values (in row order), splits them into
+        pk_values (in pk_index order) and non_pk_values (in row order).
+        """
+        assert len(all_values) == len(self.all_columns)
+
+        pk_values = [None] * len(self.pk_columns)
+        non_pk_values = []
+        for value, column in zip(all_values, self.all_columns):
+            pk_index = _pk_index(column)
+            if pk_index is None:
+                non_pk_values.append(value)
+            else:
+                pk_values[pk_index] = value
+
+        return tuple(pk_values), tuple(non_pk_values)
+
+    def principal_hash(self):
+        """
+        Returns a hash of the principal properties of this schema.
+        Less important properties can be changed without changing the hash.
+        """
+        # TODO - just hash the principal properties.
+        return _hexhash(self.dumps())
+
+
+# TODO: this class is unfinished, and so doesn't extend DatasetStructure.
+class Dataset2:
+    """
+    - Uses messagePack to serialise features.
+    - Stores each feature in a blob with path dependent on primary key values.
+    - Add at any location: `sno import GPKG:my.gpkg:mytable path/to/mylayer`
+
+    any/structure/mylayer/
+      .sno-table/
+        meta/
+          schemas/
+            current             = [current-schema-hash]
+            [schema-x-hash]     = [schema-x JSON]
+            [schema-y-hash]     = [schema-y JSON]
+            ...
+
+        [hex(pk-hash):2]/
+          [hex(pk-hash):2]/
+            [base64(pk-value)]  = [msgpack([schema-z-hash, value0, value1, ...])]
+
+    Dataset2 is initialised pointing at a particular directory tree, and uses that
+    to read features and schemas. However, it never writes to the tree, since this
+    is not straight-forward in git/sno and involves batching writes into a commit.
+    Therefore, there are no methods which write, only methods which return things
+    which *should be written*. The caller must write these to a commit.
+    """
+
+    def __init__(self, tree):
+        # Generally a pygit2.Tree, but conceptually just a directory tree
```

Comment:
ha, riffing the pathlib-style approach in pygit2 makes this quite elegant.
![](https://media3.giphy.com/media/lGV8Xw3pCyB20/giphy.gif)